### PR TITLE
FortiGate Enhancement

### PR DESCRIPTION
**Pull request by LiorQM**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
A few sentences describing the overall goals of the pull request's commits.

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/FortiGate/Integrations/FortiGate/FortiGate.py [hunk 1]
```diff
@@ -103,1209 +2141,3696 @@ def create_addr_string(list_of_addr_data_dicts):
...
+        "source6": "srcaddr6",
+        "destination6": "dstaddr6",
+        "log": "logtraffic",
+        "negate_source": "srcaddr-negate",
+        "negate_destination": "dstaddr-negate",
+        "negate_source6": "srcaddr6-negate",
+        "negate_destination6": "dstaddr6-negate",
+        "negate_service": "service-negate",
+    }
+    api_key = field_to_api_key.get(field, field)
```
Comment (omerKarkKatz) [inline Packs/FortiGate/Integrations/FortiGate/FortiGate.py:4601]: please change the api_key terminology its confusing.

File: Packs/FortiGate/Integrations/FortiGate/FortiGate.py [hunk 2]
```diff
@@ -103,1209 +2141,3696 @@ def create_addr_string(list_of_addr_data_dicts):
...
+
+    if field in {
```
Comment (omerKarkKatz) [inline Packs/FortiGate/Integrations/FortiGate/FortiGate.py:4603]: Please extract the 2 nested 'if's to a separate function with a proper doc_string explaining the purpose of this code.

**Review by omerKarkKatz — COMMENTED** [review]
Hey @LiorQM, Good job! thanks for the hard work.
Please go over my comments and fix accordingly.
Also please also go over pre-commit errors this is now a mandatory step in the internal PR. Please try to fix all errors that you dont think are supposed to fail because of old deprecated commands. 
Thanks !

File: Packs/FortiGate/Integrations/FortiGate/FortiGate.py [hunk 3]
```diff
@@ -1,93 +1,2131 @@
...
+        """Due to limited amount of simultaneous connections we log out."""
+        if Client.IS_ONLINE:
+            self._http_request(
+                method="POST",
+                full_url=urljoin(self.server, "logout"),
+                resp_type="response",
+                error_handler=Client._error_handler,
+            )
+
+    def _get_filter(self, field: str | None, value: str | None) -> str | None:
```
[the remark this answers is not in the document]
Comment (LiorQM) [reply Packs/FortiGate/Integrations/FortiGate/FortiGate.py:228]: added.

File: Packs/FortiGate/Integrations/FortiGate/FortiGate_test.py [hunk 4]
```diff
@@ -0,0 +1,1678 @@
...
+
+@pytest.mark.parametrize(
+    "args, expected_result",
+    [
+        # Successful identification of each service type
+        ({"start_ip": "0.0.0.0", "end_ip": "0.0.0.0", "tcpRange": "80"}, "TCP/UDP/SCTP"),
+        ({"ip_protocol": "6"}, "IP"),
+        ({"icmp_version": "icmp", "icmp_type": "8", "icmp_code": "0"}, "icmp"),
+        ({"icmp_version": "icmp6", "icmp_type": "128", "icmp_code": "0"}, "icmp6"),
+        # Add more cases for each type
```
[the remark this answers is not in the document]
Comment (LiorQM) [reply Packs/FortiGate/Integrations/FortiGate/FortiGate_test.py:164]: removed comment and added more cases.

File: Packs/FortiGate/Integrations/FortiGate/FortiGate_test.py [hunk 5]
```diff
@@ -0,0 +1,1678 @@
...
+        FortiGate.get_address_type(args)
+
+    assert str(exc_info.value) == error_message
+
+
+@pytest.mark.parametrize(
+    "args, expected_result",
+    [
+        # Successful identification of each service type
+        ({"start_ip": "0.0.0.0", "end_ip": "0.0.0.0", "tcpRange": "80"}, "TCP/UDP/SCTP"),
```
[the remark this answers is not in the document]
Comment (LiorQM) [reply Packs/FortiGate/Integrations/FortiGate/FortiGate_test.py:160]: added

File: Packs/FortiGate/Integrations/FortiGate/FortiGate.py [hunk 6]
```diff
@@ -103,1209 +2141,3696 @@ def create_addr_string(list_of_addr_data_dicts):
...
+        input_protocol_type = get_service_type(args)
+        api_protocol_type = result.get("protocol")
+
+        if input_protocol_type != api_protocol_type:
+            raise DemistoException(
+                f"The service '{name}' is of type '{api_protocol_type}',"
+                f" which is not compatible with the requested type '{input_protocol_type}'."
+            )
+
+    if bool(action) != any((tcp_port_ranges, udp_port_ranges, sctp_port_ranges)):
```
[the remark this answers is not in the document]
Comment (LiorQM) [reply Packs/FortiGate/Integrations/FortiGate/FortiGate.py:4144]: moved.

File: Packs/FortiGate/Integrations/FortiGate/FortiGate.py [hunk 7]
```diff
@@ -103,1209 +2141,3696 @@ def create_addr_string(list_of_addr_data_dicts):
...
+            ("tcp-portrange", "udp-portrange", "sctp-portrange"),
+        ):
+            api_port_range = set(result.get(api_key, "").split())
+
+            if action == "add":
+                port_ranges[key] = list(api_port_range.union(value))
+            else:  # action == "remove"
+                port_ranges[key] = list(api_port_range.difference(value))
+
+    validate_optional_ipv4_addresses(start_ip, end_ip)
```
[the remark this answers is not in the document]
Comment (LiorQM) [reply Packs/FortiGate/Integrations/FortiGate/FortiGate.py:4165]: moved.

File: Packs/FortiGate/Integrations/FortiGate/FortiGate.py [hunk 8]
```diff
@@ -1,93 +1,2131 @@
...
+        protocol_type: sum(1 for arg in arg_names if args.get(arg) is not None)
+        for protocol_type, arg_names in protocol_type_to_arg_names.items()
+    }
+
+    # Identify fully set, partially set, and mixed groups
+    fully_set_protocol_types = []
+    partially_set_protocol_types = []
+
+    for protocol_type, count in protocol_type_to_arg_counts.items():
+        if count >= len(protocol_type_to_arg_names[protocol_type]):
```
[the remark this answers is not in the document]
Comment (LiorQM) [reply Packs/FortiGate/Integrations/FortiGate/FortiGate.py:2040]: Not really, it is just some thing I usually do as a safety net.
I guess because all the variables are within the same scope I'll change it to only `eq`.

File: Packs/FortiGate/Integrations/FortiGate/FortiGate_test.py [hunk 9]
```diff
@@ -0,0 +1,1678 @@
...
+)
+def test_build_service_table(build_table: Callable, items: dict[str, Any], expected_table: dict[str, Any]):
+    """
+    Scenario:
+    - Test the build_service_table function's ability to correctly format data into a service table.
+    """
+    assert build_table(items) == expected_table
+
+
+def testbuild_service_group_table():
```
[the remark this answers is not in the document]
Comment (LiorQM) [reply Packs/FortiGate/Integrations/FortiGate/FortiGate_test.py:521]: added

File: Packs/FortiGate/Integrations/FortiGate/FortiGate.py [hunk 10]
```diff
@@ -103,1209 +2141,3696 @@ def create_addr_string(list_of_addr_data_dicts):
...
+    Raises:
+        DemistoException:
+            - If both versions of address were provided.
+            - If none of the versions of address were provided.
+
+    Returns:
+        CommandResults: Outputs of the command that represent an entry in warroom.
+    """
+    name = args.get("policyName", "")
+    vdom = args.get("vdom", DEFAULT_VDOM)
```
[the remark this answers is not in the document]
Comment (LiorQM) [reply Packs/FortiGate/Integrations/FortiGate/FortiGate.py:4476]: this was done to avoid any breaking changes, added comment as suggested.

File: Packs/FortiGate/Integrations/FortiGate/FortiGate.py [hunk 11]
```diff
@@ -103,1209 +2141,3696 @@ def create_addr_string(list_of_addr_data_dicts):
...
+    tcp_port_ranges = argToList(args.get("tcpRange"))
+    udp_port_ranges = argToList(args.get("udpRange"))
+    sctp_port_ranges = argToList(args.get("sctpRange"))
+    icmp_type = arg_to_number(args.get("icmp_type"))
+    icmp_code = arg_to_number(args.get("icmp_code"))
+    ip_protocol = arg_to_number(args.get("ip_protocol"))
+    action = args.get("action")
+
+    input_protocol_type = None
+    response = client.list_firewall_services(name, vdom)
```
[the remark this answers is not in the document]
Comment (LiorQM) [reply Packs/FortiGate/Integrations/FortiGate/FortiGate.py:4118]: The list call is necessary to validate the current protocol type and  maintain members in `TCP/UDP/SCTP`.
When updating a protocol in GUI it would work, but through the API an error is returned.

File: Packs/FortiGate/Integrations/FortiGate/FortiGate.py [hunk 12]
```diff
@@ -103,1209 +2141,3696 @@ def create_addr_string(list_of_addr_data_dicts):
...
+    table = []
+    header_to_key = {
+        "Name": "Name",
+        "Category": "Category",
+        "Protocol": "Protocol",
     }
 
-    contents = service_group_context
-    context['Fortigate.ServiceGroup(val.Name && val.Name === obj.Name)'] = service_group_context
+    protocol_to_handler = {
```
[the remark this answers is not in the document]
Comment (LiorQM) [reply Packs/FortiGate/Integrations/FortiGate/FortiGate.py:2372]: added

File: Packs/FortiGate/Integrations/FortiGate/FortiGate.py [hunk 13]
```diff
@@ -103,1209 +2141,3696 @@ def create_addr_string(list_of_addr_data_dicts):
...
+    name = args.get("groupName", "")
+    vdom = args.get("vdom", DEFAULT_VDOM)
+    comment = args.get("comment")
+    input_members = argToList(args.get("serviceName"))
+    action = args.get("action")
+
+    if bool(input_members) != bool(action):
+        raise DemistoException("`serviceName` must be set with `action`.")
+
+    response = client.list_firewall_service_groups(name, vdom)
```
[the remark this answers is not in the document]
Comment (LiorQM) [reply Packs/FortiGate/Integrations/FortiGate/FortiGate.py:4322]: Like any other command, if the user provides an identifier that doesn't exist then they will receive a `404 Not Found` which is reasonable.

File: Packs/FortiGate/Integrations/FortiGate/FortiGate.py [hunk 14]
```diff
@@ -103,1209 +2141,3696 @@ def create_addr_string(list_of_addr_data_dicts):
...
+    """
+    return next(iter(response.get("results", [])), {})
 
-    if policy_field in field_to_api_key:
-        policy_field = field_to_api_key[policy_field]
 
-    if policy_field in {'srcaddr', 'dstaddr'}:
-        policy_field_value = generate_src_or_dst_request_data(
-            policy_id, policy_field, policy_field_value, keep_original_data, add_or_remove)
+def handle_group_items_by_action(input_items: list[str], action: str | None, items: list[str]) -> list[str]:
```
[the remark this answers is not in the document]
Comment (LiorQM) [reply Packs/FortiGate/Integrations/FortiGate/FortiGate.py:2585]: added

File: Packs/FortiGate/Integrations/FortiGate/FortiGate.py [hunk 15]
```diff
@@ -103,1209 +2141,3696 @@ def create_addr_string(list_of_addr_data_dicts):
...
+    )
+
+    client.update_firewall_service_group(
+        name=name,
+        vdom=vdom,
+        comment=comment,
+        members=members,
+    )
+
+    response = client.list_firewall_service_groups(name, vdom)
```
[the remark this answers is not in the document]
Comment (LiorQM) [reply Packs/FortiGate/Integrations/FortiGate/FortiGate.py:4338]: In the pre-enhancement some of the commands made another request to the API to receive the updated data.
Only `GET` requests return information in this API.

File: Packs/FortiGate/Integrations/FortiGate/FortiGate.py [hunk 16]
```diff
@@ -103,1209 +2141,3696 @@ def create_addr_string(list_of_addr_data_dicts):
...
+    )
 
-    create_address_request(address_name, address, mask, fqdn)
 
-    if address:
-        address_dict = {
-            'Name': address_name,
-            'IPAddress': address
+@logger
+def policy_addr_array_from_arg(policy_addr_data, is_data_string=True):
```
[the remark this answers is not in the document]
Comment (LiorQM) [reply Packs/FortiGate/Integrations/FortiGate/FortiGate.py:5393]: removed and added new comment.

File: Packs/FortiGate/Integrations/FortiGate/FortiGate.py (hunk 1 above)
Comment (LiorQM) [reply Packs/FortiGate/Integrations/FortiGate/FortiGate.py:4601]: changed.

File: Packs/FortiGate/Integrations/FortiGate/FortiGate.py (hunk 2 above)
Comment (LiorQM) [reply Packs/FortiGate/Integrations/FortiGate/FortiGate.py:4603]: I'm not too fond of nesting, but isn't the logic already broken apart to multiple functions enough already? Also this will mean I'll have to pass a `Callable` for `client.list_firewall_policies` plus ~5 other arguments.

File: Packs/FortiGate/Integrations/FortiGate/FortiGate.py [hunk 17]
```diff
@@ -103,1209 +2141,3696 @@ def create_addr_string(list_of_addr_data_dicts):
...
+        raise DemistoException(f"The action '{action}' is only compatible with TCP/UDP/SCTP.")
+
+    port_ranges = {
+        "tcp_port_ranges": tcp_port_ranges,
+        "udp_port_ranges": udp_port_ranges,
+        "sctp_port_ranges": sctp_port_ranges,
+    }
+
+    if action:
+        for (key, value), api_key in zip(
```
[the remark this answers is not in the document]
Comment (LiorQM) [reply Packs/FortiGate/Integrations/FortiGate/FortiGate.py:4154]: extracted and added.

File: Packs/FortiGate/Integrations/FortiGate/FortiGate.py [hunk 18]
```diff
@@ -103,1209 +2141,3696 @@ def create_addr_string(list_of_addr_data_dicts):
...
+        filter_value=args.get("filter_value"),
+        format_fields=format_fields,
+    )
+
+    # Handle VDOM == *
+    responses = raw_response if isinstance(raw_response, list) else [raw_response]
+    outputs: list = []
+    found_output = None
+
+    for response in responses:
```
[the remark this answers is not in the document]
Comment (LiorQM) [reply Packs/FortiGate/Integrations/FortiGate/FortiGate.py:4408]: added

File: Packs/FortiGate/Integrations/FortiGate/FortiGate.py [hunk 19]
```diff
@@ -103,1209 +2152,3745 @@ def create_addr_string(list_of_addr_data_dicts):
...
+    source_address6 = args.get("source6", "")
+    source_addresses6 = argToList(source_address6)
+    # Keep users input for the context output to avoid breaking changes.
+    destination_address6 = args.get("destination6", "")
+    destination_addresses6 = argToList(destination_address6)
+    negate_source_address = args.get("negate_source_address")
+    negate_destination_address = args.get("negate_destination_address")
+    # Keep users input for the context output to avoid breaking changes.
+    service = args.get("service")
+    services = argToList(service)
```
Comment (omerKarkKatz) [inline Packs/FortiGate/Integrations/FortiGate/FortiGate.py:4552]:
```suggestion
    # preserve deprecated command outputs
    source_interface = args.get("sourceIntf")
    destination_interface = args.get("dstIntf")
    source_address = args.get("source", "")
    destination_address = args.get("destination", "")
    source_address6 = args.get("source6", "")
    destination_address6 = args.get("destination6", "")
    negate_source_address = args.get("negate_source_address")
    service = args.get("service")
    
    source_interfaces = argToList(source_interface)
    destination_interfaces = argToList(destination_interface)
    source_addresses = argToList(source_address)
    destination_addresses = argToList(destination_address)
    source_addresses6 = argToList(source_address6)
    destination_addresses6 = argToList(destination_address6)
    negate_destination_address = args.get("negate_destination_address")
    services = argToList(service)
```

File: Packs/FortiGate/TestPlaybooks/playbook-FortiGate_Test.yml [hunk 20]
```diff
@@ -0,0 +1,5677 @@
+id: FortiGate
```
Comment (omerKarkKatz) [inline Packs/FortiGate/TestPlaybooks/playbook-FortiGate_Test.yml:1]: If this is a new test-PB reduce to version -1

**Comment by omerKarkKatz** [comment]
Please take the old TEST-pb (little g in FortiGate delete its content and paste your new TestPB). Delete the new added file.

File: Packs/FortiGate/Integrations/FortiGate/FortiGate.py [hunk 21]
```diff
@@ -103,1209 +2152,3745 @@ def create_addr_string(list_of_addr_data_dicts):
...
+    """
+    port_ranges = {
+        "tcp_port_ranges": tcp_port_ranges,
+        "udp_port_ranges": udp_port_ranges,
+        "sctp_port_ranges": sctp_port_ranges,
     }
 
-    if policy_field in field_to_api_key:
-        policy_field = field_to_api_key[policy_field]
+    for (key, value), api_key in zip(
```
Comment (omerKarkKatz) [inline Packs/FortiGate/Integrations/FortiGate/FortiGate.py:2641]: Please change the api_key terminology.

File: Packs/FortiGate/Integrations/FortiGate/FortiGate.py [hunk 22]
```diff
@@ -103,1209 +2152,3745 @@ def create_addr_string(list_of_addr_data_dicts):
...
+        comment=comment,
+        category=category,
+        protocol_type=input_protocol_type,
+        start_ip=start_ip,
+        end_ip=end_ip,
+        fqdn=fqdn,
+        icmp_type=icmp_type,
+        icmp_code=icmp_code,
+        ip_protocol=ip_protocol,
+        **port_ranges,
```
Comment (omerKarkKatz) [inline Packs/FortiGate/Integrations/FortiGate/FortiGate.py:4249]: Call the arguments explicitly. port_ranges.get('..')

File: Packs/FortiGate/TestPlaybooks/playbook-FortiGate_Test.yml (hunk 20 above)
Comment (LiorQM) [reply Packs/FortiGate/TestPlaybooks/playbook-FortiGate_Test.yml:1]: reduced.

File: Packs/FortiGate/Integrations/FortiGate/FortiGate.py (hunk 19 above)
Comment (LiorQM) [reply Packs/FortiGate/Integrations/FortiGate/FortiGate.py:4552]: added.

File: Packs/FortiGate/Integrations/FortiGate/FortiGate.py (hunk 21 above)
Comment (LiorQM) [reply Packs/FortiGate/Integrations/FortiGate/FortiGate.py:2641]: changed

File: Packs/FortiGate/Integrations/FortiGate/FortiGate.py (hunk 22 above)
Comment (LiorQM) [reply Packs/FortiGate/Integrations/FortiGate/FortiGate.py:4249]: updated.

File: Packs/FortiGate/TestPlaybooks/playbook-Fortigate_Test.yml [hunk 23]
```diff
@@ -1,1213 +1,5674 @@
-id: Fortigate Test
+id: FortiGate
```
Comment (omerKarkKatz) [inline Packs/FortiGate/TestPlaybooks/playbook-Fortigate_Test.yml:1]: Why did you change the id?

Comment (LiorQM) [reply Packs/FortiGate/TestPlaybooks/playbook-Fortigate_Test.yml:1]: I've copied all of the other files content and replaced the current one as you asked.
I guess change was also included in it.
Would you like me to change the `name` and `id`?

Comment (LiorQM) [reply Packs/FortiGate/TestPlaybooks/playbook-Fortigate_Test.yml]: changed `id` and `name` back to the previous value.

Comment (omerKarkKatz) [reply Packs/FortiGate/TestPlaybooks/playbook-Fortigate_Test.yml]: Look at the conf.json file it already contains the Fortigate_Test PB, also your yml file 
image
Keep the original id/name/display

**Review by omerKarkKatz — APPROVED** [review]
